### DISCUSSION TITLE
Where can I find tutorials or examples of using Supabase with Prisma, but still using Supabase auth?

### DISCUSSION THREAD

**Discussion by lumenwrites** [supabase/supabase#6716] [Questions]
Hi! I'm new to Supabase. I'm building my first project with Next.js, Supabase, and Prisma.

I want to be able to use Supabase auth (to make it easier for myself to implement auth), but I want to set up most of the tables using Prisma (so that it's easier for me to manage schema and use Prisma api).

Is that a reasonable approach? Is it difficult to implement something like this? Are there any tutorials/examples I can use?

**Comment by burggraf** [comment]
Here's a good place to start:  https://supabase.com/docs/guides/integrations/prisma

**Reply by lumenwrites** [reply]
Thanks a lot for your reply!

Unfortunately it doesn't answer my question, I want to clarify - I know how to connect Prisma to Supabase, and I know how to set up Supabase auth, I just don't know how to make them work togetehr.

As far as I understand, Supabase automatically creates some sort of table for users and does some magic behind the scenes, and I can't figure out whether I'd be able to plug Prisma into that (and whether that would be a good idea).

What I'm worried about is that I'll commit to Supabase auth and API right now, but then encounter some limitations down the road, and need to go figure out how to set up my own API and write functions that do what I want. And right now, I don't know how difficult that will be to add on top of everything Supabase does automatically.

**Reply by raptor008** [reply]
I'm very new to Supabase and I'm looking at the same thing. From what I gather so far (not saying much), it seems like Prisma and Supabase Auth should be fine together. I don't think you'll want to use Prisma to modify any of the pre-installed tables anyway (e.g., anything in the "auth" schema such as "users", "refresh_tokens", etc.). All the tables you would create with Prisma would be in "public" or something else you create.

What I'm trying to figure out is if it's better to create tables that will primarily be private (meaning no client-side direct access) in the "public" schema and just turn on Row Level Security without any policies (default deny) or attempt to create a new schema called "private". It seems that app.supabase.io treats custom user-created schemas as second-class outside of the SQL editor (e.g., can't add tables with the GUI as they all seem to end up in the "public" schema, they don't show up under the "Database" tab, can't add rows in the "Table Editor"). I'm not sure if this is a permission issue that can/can't be solved or if I'm just missing something.

**Reply by burggraf** [reply]
Supabase auth will work fine with Prisma.  Auth is handles by GoTrue, and our version is open source, of course.  When you use the Supabase apis to create and edit users, you're just maintaining the `auth.users` table which you have full access to.  There should be no problems here.

Regarding adding and maintaining additional user data and handling the security for that, @raptor008 has it right.  The way I handle this is to create a public table, say `public.user_data` that has foreign key to the `auth.users` table, and I just set an RLS policy on that table based on who I want to be able to view and edit that data.  Usually I give users the ability to edit their own data (think user profile data) and everyone the ability to view that data (or, sometimes just authenticated users.)

**Reply by raptor008** [reply]
Thanks for the reply @burggraf! Not to hijack @lumenwrites thread here but: Is there any benefit to defining a custom schema and not using "public". Let's say I've got a table that I only want the server-side of my application to access. Would you just put this table in "public" and set RLS on without any policies (for default deny) vs. create a schema called, for example, "private"?

I guess where I'm coming from here is I'm using RedwoodJS and all of my DB queries are already on the server-side anyway. So the idea of a hosted database with a public API is new to me in practice. I guess my question is basically asking what is best practice for Supabase is this regard. I don't want to try to force Supabase to do something that goes against the grain so to speak.

**Reply by burggraf** [reply]
You could do it either way, really.  If you create a private schema users won't have any access to it by default, and that'll probably be the most performant way to handle things.  But then again, if you're never giving users access to the data through the api (postgrest) then you could just as easily remove all database access for the anon and authenticated roles and then just leave everything in the public schema.

**Reply by lumenwrites** [reply]
@burggraf Thank you very much for your reply!

I have learned how to use Supabase auth from [this course](https://egghead.io/lessons/supabase-use-postgres-functions-to-implement-database-logic-with-supabase), it shows how to create my own `profiles` table, and use trigger functions to automatically add a record to it every time the user is added to the `users` table managed by Supabase.

Now I want to be able to create and manage my `profiles` table using Prisma - I want to define it in my Prisma schema, and then automatically add a row to this table every time Supabase  creates a `user`, as described above.

But I don't understand, how would I be able to make a foreign key connection from the `profiles` table defined in my Prisma schema to the `users` table managed by Supabase? Since the `users` table is not defined in my Prisma schema, I won't be able to connect `profile` to the `user` by foreign key, right?

Or maybe I'm misunderstanding something. Can you give me some advice on how this should work?

**Reply by VadimZP** [reply]
@lumenwrites try to use `npx prisma db pull` after you have done what was shown in those 2 videos in the course you provided a link to? 

But don't forget to edit your schema like this befor pulling:
```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["auth", "public"]
}
```

This is the result I have got in the end:

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["auth", "public"]
}

model Review {
  id       Int     @id @default(autoincrement())
  content  String
  author   profile @relation(fields: [authorId], references: [id])
  authorId String

  @@map("reviews")
  @@schema("public")
}

/// This model contains row level security and requires additional setup for migrations. Visit https://pris.ly/d/row-level-security for more info.
model profile {
  id      String   @id(map: "profiles_pkey") @db.Uuid
  email   String?  @db.VarChar(255)
  users   users    @relation(fields: [id], references: [id], onDelete: Cascade, onUpdate: NoAction)
  reviews Review[]

  @@schema("public")
}
```

I needed to create a relationship between my `profile` and `Review` models.
But I am facing this [problem](https://github.com/prisma/prisma/issues/17734#issuecomment-1510359920) now

**Comment by burggraf** [comment]
Have Prisma write to the `public` schema, not a separate schema.

**Reply by lumenwrites** [reply]
Yes, so my `profiles` table, the one I want to define in `schema.prisma` file, is in the Supabase public schema, right?

But now I need to make a foreign key from `profiles` to `users` table (the one automatically created by Supabase).

If I was creating everything in Prisma, I'd do something like:

```
model Profile {
  id     String @id @default(cuid())
  // This is how I would make a foreign key in prisma,
  // if the User table was defined in my schema.prisma file
  user   User   @relation("UserProfile", fields: [userId], references: [id])
  userId String

  // The rest of the fields...
  // [...]
}
```

But my User table is not defined in my `prisma.schema` file, it's created by Supabase. So how do I create a foreign key from Profile to `user`?

**Reply by burggraf** [reply]
I'm not a Prisma user, but you could create your `profile` table in the Supabase dashboard, and set up the foreign key relationship to `auth.users` there.    You can use the dashboard UI, or use the SQL Query Editor in the dashboard and create the table and foreign key constraints using regular SQL commands.

**Reply by MentalGear** [reply]
@lumenwrites Maybe setup the `profiles` tables via supabase GUI, then use Prisma introspection to generate the schema?
Or did you find a way to handle it?

**Comment by alexboii** [comment]
has anyone figured this out?

**Comment by christopheragnus** [comment]
Anyone solved this yet?

**Comment by feelixe** [comment]
Prisma does not support foreign keys between schemas yet.
So there's really no point in using the auth schema from Supabase in Prisma as I see it.

I solved this by setting up Prisma like usual with my Supabase database URL.
Instead of having a foreign key to the Supabase user on my models, I had a string field.
There are some drawbacks to this like not being able to cascade delete , not being sure if the user exists delete etc.
Whenever I need to retrieve the user server side I just used the Supabase SDK with the service role key.
